File: freight-triage-polyglot-ledger-recovery/environment/native/src/stats/median_milli.cpp
#include "freight/stats.h"

#include <algorithm>

namespace freight {

// median_milli kernel.
long long stat_median_milli(const std::vector<long long>& series) {
  if (series.empty()) {
    return 0;
  }
  std::vector<long long> sorted(series);
  std::sort(sorted.begin(), sorted.end());
  size_t middle = sorted.size() / 2;
  if (sorted.size() % 2 == 1) {
    return sorted[middle] * 1000;
  }
  return statFloorDiv((sorted[middle - 1] + sorted[middle]) * 1000, 2);
}

}  // namespace freight

File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/stats/MedianMilli.java
package com.freight.stats;

/** median_milli kernel. */
public final class MedianMilli implements StatKernel {

    @Override
    public String name() {
        return "median_milli";
    }

    @Override
    public long apply(long[] series) {
        if (series.length == 0) {
            return 0L;
        }
        long[] sorted = series.clone();
        java.util.Arrays.sort(sorted);
        int middle = sorted.length / 2;
        if (sorted.length % 2 == 1) {
            return sorted[middle] * 1000L;
        }
        return StatSupport.floorDiv((sorted[middle - 1] + sorted[middle]) * 1000L, 2L);
    }
}
